src/compas_eve/core.py

```diff
@@ -58,6 +61,150 @@ def unadvertise(self, topic):
         pass
 
 
+class MessageCodec:
+    """
+    Base class for message codecs.
+
+    This class defines the interface for message encoding and decoding.
+    Subclasses should implement the `encode` and `decode` methods to handle
+    specific serialization formats.
+    """
+
+    def encode(self, message):
+        """
+        Encode a message into a serialized format.
+
+        Parameters
+        ----------
+        message : Message or dict
+            The message to encode.
+
+        Returns
+        -------
+        bytes or str
+            The encoded message.
+
+        Raises
+        ------
+        NotImplementedError
+            If the method is not implemented by a subclass.
+        """
+        raise NotImplementedError
+
+    def decode(self, message):
+        """
+        Decode a serialized message back into a `Message` object.
+
+        Parameters
+        ----------
+        message : bytes or str
+            The serialized message to decode.
+
+        Returns
+        -------
+        Message
+            The decoded message.
+
+        Raises
+        ------
+        NotImplementedError
+            If the method is not implemented by a subclass.
+        """
+        raise NotImplementedError
+
+
+class JsonMessageCodec(MessageCodec):
+    """
+    Message codec for JSON serialization.
+
+    This codec handles encoding and decoding messages using JSON format.
+    It supports messages that are instances of `Message` or dictionaries.
+    """
+
+    def encode(self, message):
+        """
+        Encode a message into a JSON string.
+
+        Parameters
+        ----------
+        message : Message or dict
+            The message to encode.
+
+        Returns
+        -------
+        str
+            The JSON-encoded message.
+        """
+        if isinstance(message, Message):
+            data = message.data
+        else:
+            data = message
```

Comment:
This logic of detecting the message type is a tad more complex (there's one more case to handle) and it should be handled centrally, so I would suggest we change `Topic._message_to_json` to `Topic._message_to_data` with the following implementation:
```
    def _message_to_data(self, message):
        """Convert a message to a data representation ready to be encoded.

        Normally, this method expects sub-classes of ``Message`` as input.
        However, it can deal with regular dictionaries as well as classes
        implementing the COMPAS data framework.
        """
        try:
            data = message.data
        except (KeyError, AttributeError):
            try:
                data = message.__data__
            except (KeyError, AttributeError):
                data = dict(message)
        return data
```

and this method can be much simpler:
```
    def encode(self, data):
        return json_dumps(data)
```

... (cont'd)